internal/backend/sessions/sessioncalls/calls.go

```diff
@@ -85,34 +117,54 @@ func (cs *calls) Call(ctx workflow.Context, params *CallParams) (sdktypes.Sessio
 
 	goCtx := temporalclient.NewWorkflowContextAsGOContext(ctx)
 
-	if params.ForceInternal || fnv.GetFunction().HasFlag(sdktypes.PureFunctionFlag) {
+	if params.ForceInternal || fnvf.HasFlag(sdktypes.PureFunctionFlag) {
 		z.Debug("function call outside activity")
 
 		var err error
 
-		if fnv.GetFunction().HasFlag(sdktypes.PrivilidgedFunctionFlag) {
+		if fnvf.HasFlag(sdktypes.PrivilidgedFunctionFlag) {
 			goCtx = sessioncontext.WithWorkflowContext(goCtx, ctx)
 		}
 
 		if _, attempt, err = cs.executeCall(goCtx, params.SessionID, seq, params.Poller, params.Executors); err != nil {
 			return sdktypes.NewSessionCallAttemptResult(sdktypes.InvalidValue, fmt.Errorf("internal call: %w", err)), nil
 		}
 	} else {
-		executorsForSessions[params.SessionID.String()] = params.Executors // HACK
+		xid := fnvf.ExecutorID()
 
-		actx := workflow.WithActivityOptions(
-			ctx,
-			workflow.ActivityOptions{
+		// neither a non-integration (like a run) or an integration that is builtin.
+		local := !xid.IsIntegrationID() || fixtures.IsBuiltinIntegrationID(xid.ToIntegrationID())
+
+		if local {
+			executorsForSessions[params.SessionID.String()] = params.Executors // HACK
+			defer func() { delete(executorsForSessions, params.SessionID.String()) }()
+		}
+
+		for retry := true; retry; {
+			aopts := workflow.ActivityOptions{
 				TaskQueue:              taskQueueName,
 				ActivityID:             fmt.Sprintf("session_call_%s_%d", params.SessionID.Value(), seq),
 				ScheduleToCloseTimeout: cs.config.Temporal.ActivityScheduleToCloseTimeout,
 				HeartbeatTimeout:       cs.config.Temporal.ActivityHeartbeatTimeout,
-			},
-		)
+			}
+
+			if local {
+				if aopts.TaskQueue = cs.uniqueWorkerCallTaskQueueName; aopts.TaskQueue == "" {
```

Comment:
Assign to `aopts` only after you check for empty string.

Reply:
not sure it'll make a more readable code. not everything is local.